### ISSUE TITLE
[Clang][RISCV] ABI lowering implementation doesn't conform to latest ABI spec about integer argument passing in stack

### ISSUE THREAD

**Issue by SixWeining** [llvm/llvm-project#57261]
As the [ABI](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc#integer-calling-convention) described:
> When passed in registers or on the stack, integer scalars narrower than XLEN bits are widened according to the sign of their type up to 32 bits, then sign-extended to XLEN bits.

But clang's implementation doesn't conform to this rule. For example:

```console
$ cat a.c
void foo(short a0, short a1, short a2, short a3, short a4, short a5, short a6, short a7, short a8);

void test() {
  foo(1, 1, 1, 1, 1, 1, 1, 1, 1);
}
```

```console
$ clang --target=riscv64 -emit-llvm -S a.c -o -
; ModuleID = 'a.c'
source_filename = "a.c"
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n64-S128"
target triple = "riscv64"

; Function Attrs: noinline nounwind optnone
define dso_local void @test() #0 {
entry:
  call void @foo(i16 noundef signext 1, i16 noundef signext 1, i16 noundef signext 1, i16 noundef signext 1, i16 noundef signext 1, i16 noundef signext 1, i16 noundef signext 1, i16 noundef signext 1, i16 noundef 1)
  ret void
}

declare dso_local void @foo(i16 noundef signext, i16 noundef signext, i16 noundef signext, i16 noundef signext, i16 noundef signext, i16 noundef signext, i16 noundef signext, i16 noundef signext, i16 noundef) #1

attributes #0 = { noinline nounwind optnone "frame-pointer"="all" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-features"="+64bit,+a,+c,+m,+relax,-save-restore" }
attributes #1 = { "frame-pointer"="all" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-features"="+64bit,+a,+c,+m,+relax,-save-restore" }

!llvm.module.flags = !{!0, !1, !2, !3}
!llvm.ident = !{!4}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 1, !"target-abi", !"lp64"}
!2 = !{i32 7, !"frame-pointer", i32 2}
!3 = !{i32 1, !"SmallDataLimit", i32 8}
!4 = !{!"clang version 16.0.0 (https://github.com/llvm/llvm-project.git 1dbe8561e7b1190da33d2e248f171816ea9d108f)"}
```

We can see the last argument which is passed in stack does not have the `signext` attribute like others.

I think this is because the rule I mention above is introduced to RISCV's ABI spec after clang's current implementation. See this: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/commit/cec39a064ee0e5b0129973fffab7e3ad1710498f

I do not find such a problem in backend, so I wonder this issue is worth fixing. How about your opinions?

**Comment by asb** [comment]
Thanks for the report, and sorry for the slightly slow response (I was out last week).

This does indeed seem to be a case where we deviate from what is now documented in the ABI and I'm kicking myself for missing the change when recently re-reviewing the ABI doc. It's unfortunate there wasn't a wider discussion about this back when it was made. I imagine we'll want to provide a command-line flag for a while so that the current lowering can be used.

**Comment by asb** [comment]
Patch to address this up at https://reviews.llvm.org/D140401 (though it's depending on some other patches that improve + refactor our ABI testing infra).

**Comment by asb** [comment]
Fixed by e0a28188d2b53427bc90cdf7f7899c64af102489
